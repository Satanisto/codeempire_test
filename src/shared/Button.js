import React from "react";
import PropTypes from "prop-types";

export default function Button({ children, onClick }) {
  return React.cloneElement(children, {
    onClick
  });
}

Button.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  onClick: PropTypes.func.isRequired
};

Button.defaultType = {
  onClick: () => {}
};
