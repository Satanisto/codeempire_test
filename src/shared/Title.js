import React from "react";
import PropTypes from "prop-types";

import styled from "styled-components";

const STitle = styled.div`
  font-size: ${props => `${props.size}${props.unit}`};
  text-transform: capitalize;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export default function Title({ text, size, unit }) {
  return (
    <STitle size={size} unit={unit}>
      {text}
    </STitle>
  );
}

Title.propTypes = {
  text: PropTypes.string.isRequired,
  size: PropTypes.number.isRequired,
  unit: PropTypes.oneOf(["rem", "px"])
};
