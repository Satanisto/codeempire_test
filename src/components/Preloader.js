import React from "react";
import PropTypes from "prop-types";

import * as S from "styles/preloader";

import { FaHamburger } from "shared/icons";
import { Appear } from "shared/animations";

export default function Preloader({ in: inProp }) {
  return (
    <Appear in={inProp}>
      <S.Preloader>
        <S.Animator>
          <FaHamburger />
        </S.Animator>
      </S.Preloader>
    </Appear>
  );
}

Preloader.propTypes = {
  in: PropTypes.bool.isRequired
};
