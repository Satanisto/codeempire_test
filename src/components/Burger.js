import React from "react";
import PropTypes from "prop-types";

import * as S from "styles/burger";
import Image from "shared/Image";
import { getImage, BURGERS } from "constants/paths";
import Title from "shared/Title";
import List from "shared/List";

export default function Burger({ item, setRef, handed, onClick }) {
  const { id, name, ingredients, bun_type, meat_type, sauce_type } = item;

  const list = ingredients
    .map(item => item.info.name)
    .concat(bun_type.info.name, meat_type.info.name, sauce_type.info.name);

  return (
    <S.Wrapper ref={setRef} handed={handed}>
      <S.Burger onClick={onClick}>
        <S.ImageWrapper>
          <Image src={getImage(BURGERS, id)} />
        </S.ImageWrapper>
        <S.TitleWrapper>
          <Title text={name} size={2} unit={"rem"} />
        </S.TitleWrapper>
        <S.ListWrapper>
          <List items={list} />
        </S.ListWrapper>
      </S.Burger>
    </S.Wrapper>
  );
}

Burger.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    ingredients: PropTypes.array.isRequired,
    bun_type: PropTypes.object.isRequired,
    meat_type: PropTypes.object.isRequired,
    sauce_type: PropTypes.object.isRequired
  }).isRequired
};
