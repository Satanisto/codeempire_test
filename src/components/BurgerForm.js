import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import { set_burger, configure_burger } from "store/actions/order";

import * as S from "styles/burger_form";

import { Appear } from "shared/animations";
import { FaClose } from "shared/icons";
import Title from "shared/Title";
import Image from "shared/Image";

import Switch from "react-switch";

import { getImage, BUNS, MEATS, SAUCES, INGREDIENTS } from "constants/paths";
import STAGES from "constants/stages";
import Button from "shared/Button";

export class BurgerForm extends Component {
  static propTypes = {
    set_burger: PropTypes.func.isRequired,
    configure_burger: PropTypes.func.isRequired,
    burger: PropTypes.shape({
      name: PropTypes.string.isRequired,
      ingredients: PropTypes.array.isRequired,
      bun_type: PropTypes.object.isRequired,
      meat_type: PropTypes.object.isRequired,
      sauce_type: PropTypes.object.isRequired
    })
  };

  constructor(props) {
    super(props);

    this.state = {
      squeeze: true,
      stage: STAGES.CONFIGURE
    };

    this.handleChangeState = this.handleChangeState.bind(this);
    this.handleChangeSqueeze = this.handleChangeSqueeze.bind(this);
    this.previousStage = this.previousStage.bind(this);
    this.nextStage = this.nextStage.bind(this);
    this.exit = this.exit.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.burger === null) {
      setTimeout(() => prevState.squeeze && this.handleChangeSqueeze(false), 1000);
    }
  }

  handleChangeState(index, newState) {
    this.props.configure_burger({
      ingredients: { [index]: { state: { $set: newState } } }
    });
  }

  handleChangeSqueeze(flag) {
    this.setState(() => ({
      squeeze: flag
    }));
  }

  getPrice() {
    const { burger } = this.props;

    return (
      burger.bun_type.info.price.uah +
      burger.meat_type.info.price.uah +
      burger.sauce_type.info.price.uah +
      burger.ingredients.reduce(
        (prev, curr) => prev + (curr.state && curr.info.price.uah),
        0
      )
    );
  }

  previousStage() {
    if (this.state.stage === STAGES.CONFIRM) {
      this.setState(() => ({
        stage: STAGES.CONFIGURE,
        squeeze: false
      }));
    }
  }

  nextStage() {
    if (this.state.stage === STAGES.CONFIGURE) {
      this.setState(() => ({
        stage: STAGES.CONFIRM,
        squeeze: true
      }));
    } else if (this.state.stage === STAGES.CONFIRM) {
      this.exit();
    }
  }

  exit() {
    this.setState(() => ({
      stage: STAGES.CONFIGURE
    }));

    this.props.set_burger(null);
  }

  render() {
    const { burger } = this.props;

    return (
      <Appear in={burger !== null}>
        <S.BurgerForm>
          <S.CloseWrapper onClick={this.exit}>
            <FaClose />
          </S.CloseWrapper>

          {burger && (
            <S.ContentWrapper>
              <Title text={burger.name} size={2} unit="rem" />
              <S.Content>
                <S.Information>
                  <S.Item>
                    {burger.bun_type.info.name} ({burger.bun_type.info.price.uah} UAH)
                  </S.Item>
                  <S.Item>
                    {burger.meat_type.info.name} ({burger.meat_type.info.price.uah} UAH)
                  </S.Item>
                  <S.Item>
                    {burger.sauce_type.info.name} ({burger.sauce_type.info.price.uah} UAH)
                  </S.Item>

                  {burger.ingredients.map((ingredient, index) => (
                    <S.Item key={ingredient.info.id}>
                      {ingredient.info.name} ({ingredient.info.price.uah} UAH){" "}
                      {this.state.stage === STAGES.CONFIGURE && (
                        <Switch
                          checked={ingredient.state}
                          onChange={() =>
                            this.handleChangeState(index, !ingredient.state)
                          }
                        />
                      )}
                    </S.Item>
                  ))}

                  <S.Price>Total: {this.getPrice()} UAH</S.Price>

                  <S.Buttons>
                    <Button
                      wrapper={S.Button}
                      onClick={this.previousStage}
                      props={{ disabled: this.state.stage === STAGES.CONFIGURE }}
                    >
                      Previous
                    </Button>

                    <Button wrapper={S.Button} onClick={this.nextStage}>
                      {this.state.stage === STAGES.CONFIGURE ? "Next" : "Confirm"}
                    </Button>
                  </S.Buttons>
                </S.Information>

                <S.Preview squeeze={this.state.squeeze}>
                  <S.BurgerItem z_index={burger.ingredients.length + 1}>
                    <Image src={getImage(BUNS, `${burger.bun_type.info.id}_top`)} />
                  </S.BurgerItem>

                  {burger.ingredients.map((ingredient, index) => (
                    <S.BurgerItem
                      key={ingredient.info.id}
                      state={!ingredient.state}
                      delete={this.state.stage === STAGES.CONFIRM && !ingredient.state}
                      z_index={burger.ingredients.length - index}
                    >
                      <Image src={getImage(INGREDIENTS, `${ingredient.info.id}`)} />
                    </S.BurgerItem>
                  ))}

                  <S.BurgerItem z_index={-1}>
                    <Image src={getImage(MEATS, `${burger.meat_type.info.id}`)} />
                  </S.BurgerItem>
                  <S.BurgerItem z_index={-2}>
                    <Image src={getImage(SAUCES, `${burger.sauce_type.info.id}`)} />
                  </S.BurgerItem>
                  <S.BurgerItem z_index={-3}>
                    <Image src={getImage(BUNS, `${burger.bun_type.info.id}_bottom`)} />
                  </S.BurgerItem>
                </S.Preview>
              </S.Content>
            </S.ContentWrapper>
          )}
        </S.BurgerForm>
      </Appear>
    );
  }
}

const mapStateToProps = state => ({
  ...state.order
});

const mapDispatchToProps = {
  set_burger,
  configure_burger
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BurgerForm);
