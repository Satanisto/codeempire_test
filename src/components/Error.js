import React from "react";
import PropTypes from "prop-types";

import { Appear } from "shared/animations";

import { Error as SError, ErrorContent as SErrorContent } from "styles/preloader";
import Title from "shared/Title";

export default function Error({ in: inProp, error }) {
  return (
    <Appear in={inProp}>
      <SError>
        <Title text="Error" size={5} unit="rem" />
        <SErrorContent>{error}</SErrorContent>
      </SError>
    </Appear>
  );
}

Error.propTypes = {
  in: PropTypes.bool.isRequired,
  error: PropTypes.any
};
