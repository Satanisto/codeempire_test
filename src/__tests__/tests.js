import React from "react";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { shallow } from "enzyme";
import Title from "shared/Title";
import { shallowToJson } from "enzyme-to-json";
import Button from "shared/Button";
import { BurgerForm } from "components/BurgerForm";
import STAGES from "constants/stages";
import partition from "shared/DensePacking/api/partitionAPI";

Enzyme.configure({ adapter: new Adapter() });

describe("Title", () => {
  it("should render correctly", () => {
    const output = shallow(<Title text={"Hello"} size={2} unit="rem" />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});

describe("Button", () => {
  it("should click", () => {
    console.log = jest.fn();
    const output = shallow(<Button onClick={() => console.log("click")}>Hello</Button>);
    output.simulate("click");
    expect(console.log).toHaveBeenCalledWith("click");
  });
});

describe("BurgerForm", () => {
  it("test state", () => {
    const test_data = `{"id":4,"name":"cheeseburger 2","ingredients":[{"state":true,"info":{"id":0,"name":"salad1","price":{"uah":32}}},{"state":true,"info":{"id":1,"name":"salad2","price":{"uah":32}}},{"state":true,"info":{"id":2,"name":"salad3","price":{"uah":32}}},{"state":true,"info":{"id":9,"name":"cheese","price":{"uah":33}}}],"bun_type":{"state":true,"info":{"id":0,"name":"brioche","price":{"uah":32}}},"meat_type":{"state":true,"info":{"id":1,"name":"salmon","price":{"uah":449}}},"sauce_type":{"state":true,"info":{"id":1,"name":"mayonnaise","price":{"uah":27}}}}`;

    const output = shallow(
      <BurgerForm
        set_burger={() => {}}
        configure_burger={() => {}}
        burger={JSON.parse(test_data)}
      />
    );

    expect(output.state().squeeze).toEqual(true);
    expect(output.state().stage).toEqual(STAGES.CONFIGURE);
    setTimeout(() => {
      expect(output.state().squeeze).toEqual(false);
      expect(output.state().stage).toEqual(STAGES.CONFIGURE);
    }, 1000);
  });
});

describe("API", () => {
  it("test partitionAPI", () => {
    const test_data = [[1, 2, 3, 4, 5, 6, 7, 8, 9], 3];
    const output = partition(...test_data);
    const correct_output = [[9, 4, 3], [8, 5, 2], [7, 6, 1]];

    expect(output).toEqual(correct_output);
  });
});
