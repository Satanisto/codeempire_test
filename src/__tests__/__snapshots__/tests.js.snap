// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title should render correctly 1`] = `
<styled.div
  size={2}
  unit="rem"
>
  Hello
</styled.div>
`;
